File: app/adcsComputeTest/airBearingDemo/Harbenger_CONTROL/Simulation - Test/Sensors.m
classdef Sensors < handle
    properties
        % Standard Deviation of Sensor Noise
        % IMU noise parameters
        sigma_gyro_x
        sigma_gyro_y
        sigma_gyro_z
        
        % Magnetometer noise parameters
        sigma_mag_x
        sigma_mag_y
        sigma_mag_z

        declination
        inclination
        field_strength
        
        % Star tracker noise parameters
        sigma_star_quaternion
        
        % Parameters
        Ts_star_tracker  % Star tracker update period
        
        % States for noise modeling
        nu_quat_w
        nu_quat_x
        nu_quat_y
        nu_quat_z
        
        % Timing
        last_star_tracker_update_t
        
        % Parameters for plotting
        last_star_tracker_update

    end
    methods
        function self = Sensors(parameters)
            % Standard Deviation of Sensor Noise
            self.sigma_gyro_x = parameters.sigma_gyro_x;
            self.sigma_gyro_y = parameters.sigma_gyro_y;
            self.sigma_gyro_z = parameters.sigma_gyro_z;
            
            self.sigma_mag_x = parameters.sigma_mag_x;
            self.sigma_mag_y = parameters.sigma_mag_y;
            self.sigma_mag_z = parameters.sigma_mag_z;

            self.declination = parameters.declination;
            self.inclination = parameters.inclination;
            self.field_strength = parameters.field_strength;

            
            self.sigma_star_quaternion = parameters.sigma_star_quaternion;
            
            % Parameters
            self.Ts_star_tracker = parameters.Ts_star_tracker;
            
            % States
            self.nu_quat_w = 0;
            self.nu_quat_x = 0;
            self.nu_quat_y = 0;
            self.nu_quat_z = 0;
            self.last_star_tracker_update_t = -Inf;
            
            % Plotting
            % Initial attitude as quaternion [w, x, y, z]
            self.last_star_tracker_update = [1; 0; 0; 0];
        end
        
        function y = update(self, states, magnetic_field, t, force_star_tracker)
            % Measure with sensors
            % states format: [qw; qx; qy; qz; wx; wy; wz]
            
            % Get gyroscope measurements
            [y_gyro_x, y_gyro_y, y_gyro_z] = self.gyroscope(states);
            
            % Get magnetometer measurements
           % [y_mag_x, y_mag_y, y_mag_z] = self.magnetometer(states, magnetic_field);
            
            % Create measurement vector
            %y = [y_gyro_x; y_gyro_y; y_gyro_z;
            %     y_mag_x; y_mag_y; y_mag_z];
            y = [y_gyro_x; y_gyro_y; y_gyro_z];
            
            % Only update the star tracker at a slower rate
            if ((t - self.last_star_tracker_update_t) >= self.Ts_star_tracker) || force_star_tracker
                [y_quat_w, y_quat_x, y_quat_y, y_quat_z] = self.star_tracker(states);
                y = [y; y_quat_w; y_quat_x; y_quat_y; y_quat_z];
                
                if ~force_star_tracker
                    self.last_star_tracker_update_t = t;
                    self.last_star_tracker_update = [y_quat_w; y_quat_x; y_quat_y; y_quat_z];
                end
            end
        end
        
        function [y_accel_x, y_accel_y, y_accel_z] = accelerometer(self, states, states_dot, ~)

            % Unpack
            qw = states(1);
            qx = states(2);
            qy = states(3);
            qz = states(4);
            wx = states(5);
            wy = states(6);
            wz = states(7);

            u_dot = states_dot(4);
            v_dot = states_dot(5);
            w_dot = states_dot(6);

           
            % Calculate the accelerometer measurments
            eta_accel_x = randn()*self.sigma_accel_x;
            eta_accel_y = randn()*self.sigma_accel_y;
            eta_accel_z = randn()*self.sigma_accel_z;

            y_accel_x= u_dot + q*w - r*v + self.g*sin(theta)+eta_accel_x;
            y_accel_y= v_dot + r*u - p*w - self.g*cos(theta)*sin(phi) + eta_accel_y;
            y_accel_z=w_dot +p*v-q*u-self.g*cos(theta)*cos(phi) + eta_accel_z;
        end
        function [y_mag_x, y_mag_y, y_mag_z] = magnetometer(self, states,~,~)
            % Unpack
            qw = states(1);
            qx = states(2);
            qy = states(3);
            qz = states(4);

           
            % Rotation Vehicle to Body
            Rv2b = 0;


            % Get declination and inclination at the location we are at
            delta = self.declination;
            i = self.inclination;

            %Magnetic feild in the body frame
            m_i = [cos(delta), -sin(delta), 0;
                    sin(delta), cos(delta), 0;
                    0, 0, 1]...
                    * [cos(-i), 0, sin(-i);
                      0, 1, 0;
                      -sin(-i), 0, cos(-i)] * [self.field_strength; 0; 0];


            % Random Noise
            eta_mag_x = randn()*self.sigma_mag_x;
            eta_mag_y = randn()*self.sigma_mag_y;
            eta_mag_z = randn()*self.sigma_mag_z;

            beta_mag = [self.beta_mag_x; self.beta_mag_y; self.beta_mag_z];
            eta_mag = [eta_mag_x; eta_mag_y; eta_mag_z];

            y_mag = Rv2b*(m_i + beta_mag) + eta_mag ;


            y_mag_x = y_mag(1);
            y_mag_y = y_mag(2);
            y_mag_z = y_mag(3);



        end

        function [y_gyro_x, y_gyro_y, y_gyro_z] = gyroscope(self, states)
            % Unpack states
            wx = states(5);
            wy = states(6);
            wz = states(7);
            
            % Calculate gyroscope measurements (with noise)
            eta_gyro_x = randn() * self.sigma_gyro_x;
            eta_gyro_y = randn() * self.sigma_gyro_y;
            eta_gyro_z = randn() * self.sigma_gyro_z;
            
            y_gyro_x = wx + eta_gyro_x;
            y_gyro_y = wy + eta_gyro_y;
            y_gyro_z = wz + eta_gyro_z;
        end
        function [y_quat_w, y_quat_x, y_quat_y, y_quat_z] = star_tracker(self, states)
           % Unpack states (quaternion part)
            qw = states(1);
            qx = states(2);
            qy = states(3);
            qz = states(4);
            
            % Calculate star tracker measurements (with noise)
            % For quaternions, noise can't be simply added as it would break the unit quaternion constraint
            % So we create a small random rotation quaternion and apply it
            
            % Create a small random rotation
            eta_angle = randn() * self.sigma_star_quaternion; % Random small angle
            eta_axis = randn(3,1);
            eta_axis = eta_axis / norm(eta_axis); % Normalize to unit vector
            
            eta_quat = [cos(eta_angle/2);
                        sin(eta_angle/2) * eta_axis];
            
            % Apply the noise as a quaternion multiplication
            quat_noisy = quatmultiply([qw; qx; qy; qz]', eta_quat');
            
            % Ensure it's a unit quaternion
            quat_norm = norm(quat_noisy);
            quat_noisy = quat_noisy / quat_norm;
            
            y_quat_w = quat_noisy(1);
            y_quat_x = quat_noisy(2);
            y_quat_y = quat_noisy(3);
            y_quat_z = quat_noisy(4);
        end
    end
end

File: app/adcsComputeTest/airBearingDemo/Harbenger_CONTROL/RustControl/src/sensors.rs
use nalgebra as na;
use rand::distributions::{Distribution, Normal};
use crate::parameters::Parameters;
use crate::utils::quaternion;

pub struct Sensors {
    // IMU noise parameters
    sigma_gyro_x: f64,
    sigma_gyro_y: f64,
    sigma_gyro_z: f64,
    
    // Magnetometer noise parameters
    sigma_mag_x: f64,
    sigma_mag_y: f64,
    sigma_mag_z: f64,
    
    // Magnetic field parameters
    declination: f64,
    inclination: f64,
    field_strength: f64,
    
    // Star tracker noise parameter
    sigma_star_quaternion: f64,
    
    // Timing parameters
    ts_star_tracker: f64,
    last_star_tracker_update_t: f64,
    
    // Last star tracker measurement (for plotting)
    last_star_tracker_update: na::Vector4<f64>,
    
    // Random number generators
    rng_gyro_x: Normal<f64>,
    rng_gyro_y: Normal<f64>,
    rng_gyro_z: Normal<f64>,
    rng_star: Normal<f64>,
}

impl Sensors {
    pub fn new(parameters: &Parameters) -> Self {
        Sensors {
            // IMU noise parameters
            sigma_gyro_x: parameters.sigma_gyro_x,
            sigma_gyro_y: parameters.sigma_gyro_y,
            sigma_gyro_z: parameters.sigma_gyro_z,
            
            // Magnetometer noise parameters
            sigma_mag_x: parameters.sigma_mag_x,
            sigma_mag_y: parameters.sigma_mag_y,
            sigma_mag_z: parameters.sigma_mag_z,
            
            // Magnetic field parameters
            declination: parameters.declination,
            inclination: parameters.inclination,
            field_strength: parameters.field_strength,
            
            // Star tracker noise
            sigma_star_quaternion: parameters.sigma_star_quaternion,
            
            // Timing
            ts_star_tracker: parameters.ts_star_tracker,
            last_star_tracker_update_t: f64::NEG_INFINITY,
            
            // Last measurement
            last_star_tracker_update: na::Vector4::new(1.0, 0.0, 0.0, 0.0),
            
            // Initialize random number generators
            rng_gyro_x: Normal::new(0.0, parameters.sigma_gyro_x),
            rng_gyro_y: Normal::new(0.0, parameters.sigma_gyro_y),
            rng_gyro_z: Normal::new(0.0, parameters.sigma_gyro_z),
            rng_star: Normal::new(0.0, parameters.sigma_star_quaternion),
        }
    }
    
    pub fn update(
        &mut self, 
        states: &na::Vector7<f64>, 
        _magnetic_field: f64, 
        t: f64,
        force_star_tracker: bool
    ) -> na::DVector<f64> {
        // Get gyroscope measurements
        let (gyro_x, gyro_y, gyro_z) = self.gyroscope(states);
        
        // Initialize measurement vector with gyroscope readings
        let mut y = na::DVector::from_vec(vec![gyro_x, gyro_y, gyro_z]);
        
        // Add star tracker measurements if it's time or forced
        if (t - self.last_star_tracker_update_t) >= self.ts_star_tracker || force_star_tracker {
            let (quat_w, quat_x, quat_y, quat_z) = self.star_tracker(states);
            
            // Add quaternion measurements to the vector
            y = na::DVector::from_vec(vec![
                gyro_x, gyro_y, gyro_z, 
                quat_w, quat_x, quat_y, quat_z
            ]);
            
            // Update last measurement time if not forced
            if !force_star_tracker {
                self.last_star_tracker_update_t = t;
                self.last_star_tracker_update = na::Vector4::new(quat_w, quat_x, quat_y, quat_z);
            }
        }
        
        y
    }
    
    fn gyroscope(&mut self, states: &na::Vector7<f64>) -> (f64, f64, f64) {
        // Extract angular velocity
        let wx = states[4];
        let wy = states[5];
        let wz = states[6];
        
        // Generate random noise
        let eta_gyro_x = self.rng_gyro_x.sample(&mut rand::thread_rng());
        let eta_gyro_y = self.rng_gyro_y.sample(&mut rand::thread_rng());
        let eta_gyro_z = self.rng_gyro_z.sample(&mut rand::thread_rng());
        
        // Add noise to measurements
        let y_gyro_x = wx + eta_gyro_x;
        let y_gyro_y = wy + eta_gyro_y;
        let y_gyro_z = wz + eta_gyro_z;
        
        (y_gyro_x, y_gyro_y, y_gyro_z)
    }
    
    fn star_tracker(&mut self, states: &na::Vector7<f64>) -> (f64, f64, f64, f64) {
        // Extract quaternion
        let qw = states[0];
        let qx = states[1];
        let qy = states[2];
        let qz = states[3];
        
        // Create a small random rotation quaternion to model star tracker noise
        let eta_angle = self.rng_star.sample(&mut rand::thread_rng());
        
        // Generate random axis
        let mut rng = rand::thread_rng();
        let eta_axis_x = self.rng_star.sample(&mut rng);
        let eta_axis_y = self.rng_star.sample(&mut rng);
        let eta_axis_z = self.rng_star.sample(&mut rng);
        
        // Normalize the axis
        let axis_norm = (eta_axis_x * eta_axis_x + 
                         eta_axis_y * eta_axis_y + 
                         eta_axis_z * eta_axis_z).sqrt();
        
        let eta_axis = na::Vector3::new(
            eta_axis_x / axis_norm,
            eta_axis_y / axis_norm,
            eta_axis_z / axis_norm
        );
        
        // Create noise quaternion
        let eta_quat = na::Vector4::new(
            (eta_angle / 2.0).cos(),
            (eta_angle / 2.0).sin() * eta_axis[0],
            (eta_angle / 2.0).sin() * eta_axis[1],
            (eta_angle / 2.0).sin() * eta_axis[2]
        );
        
        // Apply noise via quaternion multiplication
        let true_quat = na::Vector4::new(qw, qx, qy, qz);
        let noisy_quat = quaternion::quatmultiply(&true_quat, &eta_quat);
        
        // Normalize
        let quat_norm = noisy_quat.norm();
        let normalized_quat = noisy_quat / quat_norm;
        
        (normalized_quat[0], normalized_quat[1], normalized_quat[2], normalized_quat[3])
    }
}